### ISSUE TITLE
hypersistence-utils-hibernate-62/63:3.7.0/1 causes random problems with Hibernate 6.4.0

### ISSUE THREAD

**Issue by atamanroman** [vladmihalcea/hypersistence-utils#696]
We upgraded a bunch of services from Quarkus 3.6.6 to 3.7.0 which in turn updated Hibernate 6.2.19 to 6.4.2. After this update we saw a few weird but reproducible Hibernate issues like 

- internal `java.lang.ClassCastException: class [Ljava.util.UUID; cannot be cast to class [B` when Hibernate resolved lazy properties because the type of the `UUID[]` parameter was somehow `com.vladmihalcea.hibernate.type.array.UUIDArrayType` instead of `org.hibernate.type.BasicArrayType`
  ```
  java.lang.ClassCastException: class [Ljava.util.UUID; cannot be cast to class [B ([Ljava.util.UUID; and [B are in module java.base of loader 'bootstrap')
    at org.hibernate.type.descriptor.jdbc.VarbinaryJdbcType$1.doBind(VarbinaryJdbcType.java:100)
    at org.hibernate.type.descriptor.jdbc.BasicBinder.bind(BasicBinder.java:61)
    at org.hibernate.sql.exec.internal.AbstractJdbcParameter.bindParameterValue(AbstractJdbcParameter.java:131)
    at org.hibernate.sql.exec.internal.AbstractJdbcParameter.bindParameterValue(AbstractJdbcParameter.java:102)
    at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.bindParameters(DeferredResultSetAccess.java:203)
    at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.executeQuery(DeferredResultSetAccess.java:233)
    at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.getResultSet(DeferredResultSetAccess.java:167)
    at org.hibernate.sql.results.jdbc.internal.JdbcValuesResultSetImpl.advanceNext(JdbcValuesResultSetImpl.java:218)
    at org.hibernate.sql.results.jdbc.internal.JdbcValuesResultSetImpl.processNext(JdbcValuesResultSetImpl.java:98)
    at org.hibernate.sql.results.jdbc.internal.AbstractJdbcValues.next(AbstractJdbcValues.java:19)
    at org.hibernate.sql.results.internal.RowProcessingStateStandardImpl.next(RowProcessingStateStandardImpl.java:66)
    at org.hibernate.sql.results.spi.ListResultsConsumer.consume(ListResultsConsumer.java:181)
    at org.hibernate.sql.results.spi.ListResultsConsumer.consume(ListResultsConsumer.java:33)
    at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.doExecuteQuery(JdbcSelectExecutorStandardImpl.java:209)
    at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.executeQuery(JdbcSelectExecutorStandardImpl.java:83)
    at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:76)
    at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:65)
    at org.hibernate.loader.ast.internal.CollectionBatchLoaderArrayParam.initializeKeys(CollectionBatchLoaderArrayParam.java:204)
    at org.hibernate.loader.ast.internal.AbstractCollectionBatchLoader.load(AbstractCollectionBatchLoader.java:94)
    at org.hibernate.loader.ast.internal.CollectionBatchLoaderArrayParam.load(CollectionBatchLoaderArrayParam.java:118)
    at org.hibernate.persister.collection.AbstractCollectionPersister.initialize(AbstractCollectionPersister.java:680)
    at org.hibernate.event.internal.DefaultInitializeCollectionEventListener.onInitializeCollection(DefaultInitializeCollectionEventListener.java:67)
    at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
    at org.hibernate.internal.SessionImpl.initializeCollection(SessionImpl.java:1692)
    at org.hibernate.collection.spi.AbstractPersistentCollection.lambda$initialize$3(AbstractPersistentCollection.java:615)
    at org.hibernate.collection.spi.AbstractPersistentCollection.withTemporarySessionIfNeeded(AbstractPersistentCollection.java:264)
    at org.hibernate.collection.spi.AbstractPersistentCollection.initialize(AbstractPersistentCollection.java:613)
    at org.hibernate.collection.spi.AbstractPersistentCollection.read(AbstractPersistentCollection.java:136)
    at org.hibernate.collection.spi.PersistentBag.iterator(PersistentBag.java:371)
    at org.hibernate.type.CollectionType.getElementsIterator(CollectionType.java:244)
    at org.hibernate.engine.spi.CascadingActions.getAllElementsIterator(CascadingActions.java:477)
    at org.hibernate.engine.spi.CascadingActions$1.getCascadableChildrenIterator(CascadingActions.java:70)
    at org.hibernate.engine.internal.Cascade.cascadeCollectionElements(Cascade.java:549)
    at org.hibernate.engine.internal.Cascade.cascadeCollection(Cascade.java:481)
    at org.hibernate.engine.internal.Cascade.cascadeAssociation(Cascade.java:442)
    at org.hibernate.engine.internal.Cascade.cascadeProperty(Cascade.java:224)
    at org.hibernate.engine.internal.Cascade.cascade(Cascade.java:157)
    at org.hibernate.event.internal.DefaultDeleteEventListener.cascadeBeforeDelete(DefaultDeleteEventListener.java:506)
    at org.hibernate.event.internal.DefaultDeleteEventListener.deleteEntity(DefaultDeleteEventListener.java:395)
    at org.hibernate.event.internal.DefaultDeleteEventListener.delete(DefaultDeleteEventListener.java:245)
    at org.hibernate.event.internal.DefaultDeleteEventListener.deletePersistentInstance(DefaultDeleteEventListener.java:221)
    at org.hibernate.event.internal.DefaultDeleteEventListener.delete(DefaultDeleteEventListener.java:161)
    at org.hibernate.event.internal.DefaultDeleteEventListener.onDelete(DefaultDeleteEventListener.java:97)
    at org.hibernate.event.internal.DefaultDeleteEventListener.onDelete(DefaultDeleteEventListener.java:84)
    at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
    at org.hibernate.internal.SessionImpl.fireDelete(SessionImpl.java:960)
    at org.hibernate.internal.SessionImpl.delete(SessionImpl.java:891)
    at org.hibernate.internal.SessionImpl.remove(SessionImpl.java:2377)
    at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.remove(TransactionScopedSession.java:167)
    at org.hibernate.engine.spi.SessionLazyDelegator.remove(SessionLazyDelegator.java:351)
    at org.hibernate.Session_OpdLahisOZ9nWRPXMsEFQmQU03A_Synthetic_ClientProxy.remove(Unknown Source)
          [...]
  ```

- `java.lang.IllegalStateException: Illegal pop() with non-matching JdbcValuesSourceProcessingState` when calling `getResultList()` but `getResultStream()` worked fine
  ```
  java.lang.IllegalStateException: Illegal pop() with non-matching JdbcValuesSourceProcessingState
          at org.hibernate.sql.results.spi.LoadContexts.deregister(LoadContexts.java:45)
          at org.hibernate.sql.results.spi.ListResultsConsumer.consume(ListResultsConsumer.java:215)
          at org.hibernate.sql.results.spi.ListResultsConsumer.consume(ListResultsConsumer.java:33)
          at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.doExecuteQuery(JdbcSelectExecutorStandardImpl.java:209)
          at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.executeQuery(JdbcSelectExecutorStandardImpl.java:83)
          at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:76)
          at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:65)
          at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.lambda$new$2(ConcreteSqmSelectQueryPlan.java:137)
          at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.withCacheableSqmInterpretation(ConcreteSqmSelectQueryPlan.java:359)
          at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.performList(ConcreteSqmSelectQueryPlan.java:300)
          at org.hibernate.query.sqm.internal.QuerySqmImpl.doList(QuerySqmImpl.java:509)
          at org.hibernate.query.spi.AbstractSelectionQuery.list(AbstractSelectionQuery.java:427)
          at org.hibernate.query.Query.getResultList(Query.java:120)
          [...]
  ```

In all cases so far, this was resolved by removing hypersistence-utils (we only used `@Type(JsonType.class)` and replaced it with `@JdbcTypeCode(SqlTypes.JSON)`). I decided to create this ticket here although I haven't been able to find the root cause yet. Maybe it saves somebody a day of debugging. As of now, I can only tell that having _hypersistence-utils-hibernate-62/63:3.7.0/1_ on the classpath broke Hibernate 6.4.0 in very random and weird ways in our setup.

Update: just tested the same cases with hypersistence-utils-hibernate-63:3.7.1 but the problems persist.

**Comment by vladmihalcea** [comment]
Try to upgrade to `3.7.1` and see if it works.

There was a problem with arrays in 3.7.0, so maybe that will fix your issue.

**Comment by aleshchynskyi** [comment]
We have the same issue (#1), and upgrading to 3.7.1 doesn't fix it. 
@atamanroman is it working for you?

**Comment by atamanroman** [comment]
@aleshchynskyi nope, updating to _-63_ and/or _3.7.1_ didn't fix the issue.

**Comment by vladmihalcea** [comment]
@atamanroman First, you will need to create a replicating test case in this repository. After you have the test case, you need to investigate what's causing it and provide a Pull Request with the fix.

**Comment by atamanroman** [comment]
@vladmihalcea I don't need to do anything (same goes for you, of course ;)). I don't expect free support here.

As I said, I opened this issue mostly for documentation so that others maybe save some of the hours I spent investigating the issue. We removed all hypersistence-utils-hibernate dependencies. I'll try to provide a reproducer, anyway.

Regarding the wrong package name: my bad, I googled the class and copied the old package.
